R/testGeno.R

```diff
@@ -1,215 +1,325 @@
-.testSingleVarMultAlleles <- function(Gtilde, Ytilde, n, k){
-    v <- ncol(Gtilde)
-    
-    var.names <- colnames(Gtilde)
-    
-    res <- matrix(NA, nrow = 1, ncol = length(var.names)*2 + 2,
-                  dimnames = list(NULL, 
-                                  c(paste0("Est.", var.names), paste0("SE.", var.names), "Joint.Stat", "Joint.Pval" ) ))
-    
-    
-    GPG <- crossprod(Gtilde)
-    GPGinv <- tryCatch(chol2inv(chol(GPG)), error = function(e) {TRUE})
-    
-    if (is.logical(GPGinv)) return(list(res = res, allelesCovMat = NA))
-    
-    GPY <- crossprod(Gtilde, Ytilde)
-    betas <- crossprod(GPGinv, GPY) ## effect estimates of the various alleles
-    res[1, grep("^Est\\.G", colnames(res))] <- betas
-    
-    sY2 <- sum(Ytilde^2)
-    RSS <- as.numeric((sY2 - crossprod(GPY, betas))/(n - k - v))
-    Vbetas <- GPGinv * RSS
-    
-    res[1, grep("^SE\\.G", colnames(res))] <- sqrt(diag(Vbetas))
-    
-    res[1, "Joint.Stat"] <- tryCatch(crossprod(betas,
-                                               crossprod(GPG, betas))/RSS, 
-                                     error = function(e) { NA })
-    
-    res[,"Joint.pval"] <- pchisq(res[,"Joint.Stat"], df = v, lower.tail = FALSE)
-    
-    res <- as.data.frame(res)
-    return(list(res = res, allelesCovMat = Vbetas))
-}
-
-
-
-
-
+
+## function that gets an n\times p matrix of p genotypes of n individuals, and a null model, and tests the genotypes associations with the outcomes. 
+## Genetic data are always assumed complete. 
+## Types of tests: 
+## Single variant: Wald, score, BinomiRare, interaction. 
+## Variant set: SKAT, burden, SKAT-O. Multiple types of p-values. Default: Davis with Koenen if does not converge. 
+
+
+# E an environmental variable for optional GxE interaction analysis. 
+testGenoSingleVar <- function(nullmod, G, E = NULL, test = c("Score", "Wald", "SAIGE", "BinomiRare", "CMP"),
+                              GxE.return.cov = FALSE, calc.score=FALSE){
+    test <- match.arg(test)
+
+    G <- .genoAsMatrix(nullmod, G)
+
+    if (test == "Wald" & nullmod$family$family != "gaussian"){
+    	test <- "Score"
+    	message("Cannot use Wald test for non-guassian families, using the score test instead.")
+    }
+    
+    if (test == "Wald" & is.null(E)){
+        Gtilde <- calcGtilde(nullmod, G)
+        res <- .testGenoSingleVarWald(Gtilde, nullmod$Ytilde,
+                                      n=length(nullmod$Ytilde), k=ncol(nullmod$model.matrix))
+    }
+    
+    if (test == "Score" & !is.null(E)){
+    	test <- "Wald"
+    	message("Cannot use Score test for GxE, using the Wald test instead.")
+    }
+        
+    if (test == "Wald" & !is.null(E)){
+        res <- .testGenoSingleVarWaldGxE(nullmod, G, E, GxE.return.cov.mat=GxE.return.cov)
+    }
+    
+    if (test == "Score"){
+        Gtilde <- calcGtilde(nullmod, G)
+        res <- .testGenoSingleVarScore(Gtilde, G, nullmod$resid)
+    }
+    
+    

```

Comment:
SAIGE should only be used for binary outcomes, right? we need to add a check for that if it's correct